go/vt/vttablet/tabletmanager/vreplication/vcopier_copy_all.go

```diff
@@ -0,0 +1,297 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vreplication
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	"google.golang.org/protobuf/encoding/prototext"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+/*
+This file is similar to vcopier.go: it handles the copy phase for the AtomicCopy where all tables
+are streamed in a single phase.
+*/
+
+type copyAllState struct {
+	vc               *vcopier
+	plan             *ReplicatorPlan
+	currentTableName string
+	tables           map[string]bool
+}
+
+// newCopyAllState creates the required table plans and sets up the copy state for all tables in the source.
+func newCopyAllState(vc *vcopier) (*copyAllState, error) {
+	state := &copyAllState{
+		vc: vc,
+	}
+	plan, err := buildReplicatorPlan(vc.vr.source, vc.vr.colInfoMap, nil, vc.vr.stats)
+	if err != nil {
+		return nil, err
+	}
+	state.plan = plan
+	state.tables = make(map[string]bool, len(plan.TargetTables))
+	for _, table := range plan.TargetTables {
+		state.tables[table.TargetName] = false
+	}
+	return state, nil
+}
+
+// copyAll copies all tables from the source to the target sequentially, finishing one table first and then moving to the next..
+func (vc *vcopier) copyAll(ctx context.Context, settings binlogplayer.VRSettings) error {
+	var err error
+
+	log.Infof("Starting copyAll for %s", settings.WorkflowName)
+	defer log.Infof("Returning from copyAll for %s", settings.WorkflowName)
+	defer vc.vr.dbClient.Rollback()
+
+	state, err := newCopyAllState(vc)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, copyPhaseDuration)
+	defer cancel()
+
+	rowsCopiedTicker := time.NewTicker(rowsCopiedUpdateInterval)
+	defer rowsCopiedTicker.Stop()
+	copyStateGCTicker := time.NewTicker(copyStateGCInterval)
+	defer copyStateGCTicker.Stop()
+
+	parallelism := getInsertParallelism()
+	copyWorkerFactory := vc.newCopyWorkerFactory(parallelism)
+	var copyWorkQueue *vcopierCopyWorkQueue
+
+	// Allocate a result channel to collect results from tasks.
+	resultCh := make(chan *vcopierCopyTaskResult, parallelism*4)
```

Comment:
explain why `parallelism*4`?

Reply:
Added comment/constant to explain. It is essentially adds a buffer for the results from the parallel copy tasks, so tasks are not stalled writing on the results channel in case the consumer of that channel is slow for some reason.